src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -170,26 +177,29 @@ library GPv2Encoding {
             s := calldataload(add(encodedTrade.offset, 172))
         }
 
-        trade.order.typeHash = ORDER_TYPE_HASH;
         trade.order.sellToken = tokens[sellTokenIndex];
         trade.order.buyToken = tokens[buyTokenIndex];
-        trade.order.kind = OrderKind(flags & 0x1);
-        trade.order.partiallyFillable = (flags >> 1) & 0x01 == 0x01;
+        trade.order.kind = OrderKind(flags & 0x01);
+        trade.order.partiallyFillable = flags & 0x02 != 0;
 
         trade.sellTokenIndex = sellTokenIndex;
         trade.buyTokenIndex = buyTokenIndex;
 
-        // NOTE: In order to avoid memory allocation and copying of the order
-        // parameters, use the memory region reserved by the caller for the
-        // order result to effeciently compute the hash in place. Furthermore,
-        // structs are not packed in Solidity, and there are 10 [`Order`] fields
-        // to hash for a total of `10 * sizeof(uint) = 320` bytes. This digest
-        // is a EIP-712 struct hash for the [`Order`] parameters.
+        // NOTE: Compute the EIP-712 order struct hash in place. The hash is
+        // computed from the order type hash concatenated with the ABI encoded
+        // order fields for a total of `10 * sizeof(uint) = 320` bytes.
+        // Fortunately, since Solidity memory structs **are not** packed, they
+        // are already laid out in memory exactly as is needed to compute the
+        // struct hash, just requiring the order type hash to be temporarily
+        // writen to the memory slot coming right before the order data.
```

Comment:
What does the slot immediately before the order data contain? Before we were writing over a useless memory slot of the order, and now?

Reply:
It can be anywhere. Note that we restore what was there before after we are done hashing now.

Reply:
We can assume there is something before because Solidity reserves some memory for stuff of size 4 words at the start. I imagine `keccak256` does not write nor uses in any way this scratch area during its execution because it's likely implemented as an opcode. Ok, looks reasonable. Does it make sense to add this to the comment?

Reply:
>  I imagine keccak256 does not write nor uses in any way this scratch area during its execution because it's likely implemented as an opcode

Yes, its the `SHA3` opcode

> We can assume there is something before because Solidity reserves some memory for stuff of size 4 words at the start

If not, we would revert. (we would write to `0xfff...ffff` which would run out of gas.